### PR TITLE
Statsd

### PR DESCRIPTION
**Pull request by jjonek**
Added AFM metrics reporting to statsd  and made some generic cleaning in the sources.

### CODE REVIEW

File: slamon_afm/tests/models_tests.py [hunk 1]
```diff
@@ -160,6 +162,209 @@ def testDontClaimErrorTasks(self):
...
+        incr_mock.assert_called_once_with('tasks.posted')
+        timing_mock.assert_not_called()
+        gauge_mock.assert_has_calls([
+            call('tasks.task-type-1.1.queue', 1),
+            call('tasks.task-type-1.1.processing', 0)
+        ], any_order=True)
+
+    def testIncDeletedStats(self):
+        task = self._addTask()
+        with patch.object(statsd, 'incr', return_value=None) as incr_mock, \
```
Comment (kryptidikettu) [inline slamon_afm/tests/models_tests.py:211]: Why are these here and in the next tests as `with`-clause and on the previous test they were decorators? I think the decorator approach is more "pretty"... I guess its more a matter of taste, but I vote for consistency above all: So either decorators _or_ `with`-clause. Not both.

File: slamon_afm/tests/models_tests.py [hunk 2]
```diff
@@ -160,6 +162,209 @@ def testDontClaimErrorTasks(self):
...
+            db.session.flush()
+
+            incr_mock.assert_called_once_with('tasks.deleted')
+            timing_mock.assert_called_once_with('tasks.task-type-1.1.deleted', ANY)
+
+    def testIncClaimedStats(self):
+        self._addTask()
+
+        # create list to iterate the generator returned by claim_tasks
+        with patch.object(statsd, 'incr', return_value=None) as incr_mock, \
```
Comment (StealthyLoner) [inline slamon_afm/tests/models_tests.py]: And here

File: slamon_afm/tests/models_tests.py (hunk 1 above)
Comment (jjonek) [inline slamon_afm/tests/models_tests.py:211]: The first one is checking if the mocks are correctly called then a task is added by the _addTask(), whereas the others want to ignore these calls

File: slamon_afm/routes/agent_routes.py [hunk 3]
```diff
@@ -192,25 +196,19 @@ def post_tasks():
...
-        if 'task_data' in data:
-            task.result_data = data['task_data']
-            task.completed = datetime.utcnow()
-            result = task.result_data
-        elif 'task_error' in data:
-            task.error = data['task_error']
-            task.failed = datetime.utcnow()
-            result = task.error
+        task.complete(
+            result_data=data['task_data'] if 'task_data' in data else None,
```
Comment (kryptidikettu) [inline slamon_afm/routes/agent_routes.py]: Isn't this the exact same thing as `result_data=data.get('task_data'), error=data.get('task_error')`?

Comment (kryptidikettu) [inline slamon_afm/routes/agent_routes.py]:
```
In [1]: dict.get?
Type:        method_descriptor
String form: <method 'get' of 'dict' objects>
Namespace:   Python builtin
Docstring:   D.get(k[,d]) -> D[k] if k in D, else d.  d defaults to None.
```

File: slamon_afm/tests/models_tests.py (hunk 1 above)
Comment (kryptidikettu) [inline slamon_afm/tests/models_tests.py:211]: A-ha, gotcha!

File: slamon_afm/tests/models_tests.py [hunk 4]
```diff
@@ -160,6 +162,209 @@ def testDontClaimErrorTasks(self):
...
+                patch.object(statsd, 'timing') as timing_mock,\
+                patch.object(statsd, 'gauge') as gauge_mock:
+            _ = list(Task.claim_tasks(self.agent, 1))
```
Comment (kryptidikettu) [inline slamon_afm/tests/models_tests.py]: What is this `_` variable? I don't see a reference to it anywhere after this... You certain this assignment is necessary?

File: slamon_afm/routes/agent_routes.py (hunk 3 above)
Comment (jjonek) [inline slamon_afm/routes/agent_routes.py]: I'm learning new things :see_no_evil: 
I'll definitely update it to make it more readable

File: slamon_afm/tests/models_tests.py (hunk 4 above)
Comment (jjonek) [inline slamon_afm/tests/models_tests.py]: Old habit I guess, from compiled languages to prevent compiler from optimizing it out. Also the related comment seems to be bit off

File: slamon_afm/models.py [hunk 5]
```diff
@@ -7,10 +8,24 @@
...
 
 db = SQLAlchemy()
 
+from slamon_afm.stats import statsd
+
+
+class TaskException(Exception):
+    task = None
+
+    def __init__(self, *args, task=None):
```
Comment (kryptidikettu) [inline slamon_afm/models.py]: To silence the flake8/pep8/pylint error "invalid syntax" (Which this most likely is), you need to change this line to:

``` python
def __init__(self, task=None, *args):
```

Comment (kryptidikettu) [inline slamon_afm/models.py]: Actually, for some reason, it breaks the tests (It silences the linter warning tho)

Comment (kryptidikettu) [inline slamon_afm/models.py]: This is what breaks with the syntax I provided

```
  File ".../slamon-agent-fleet-manager/slamon_afm/models.py", line 249, in complete
    raise IllegalStateException("Cannot complete a tasks that is already completed!", task=self)
TypeError: __init__() got multiple values for argument 'task'
```

Comment (kryptidikettu) [inline slamon_afm/models.py]: Is it really necessary to get the `*args` anyways and then pass it to the `Exception` class? Python tutorials do not show this kind of behavior: https://docs.python.org/3.4/tutorial/errors.html#user-defined-exceptions

According to that page, the correct way would be:

``` python
class TaskException(Exception):
    def __init__(self, task=None):
        self.task = task
```

Edit: The error message needs the `*args`

Comment (jjonek) [inline slamon_afm/models.py]: the `__str__` method of the Exception class will return all values passed to `__init__`.. The only motivation in the current solution is to pass readable message for the base Exeception, which could be also achieved by explicitly passing e.g. `msg` arg to the super.

Comment (kryptidikettu) [inline slamon_afm/models.py]: I managed to silence the linter warnings and then get a correct behavior with this:

``` python
class TaskException(Exception):
    def __init__(self, *args, **kwargs):
        super(TaskException, self).__init__(*args)
        self.task = kwargs.get('task')
```

Comment (kryptidikettu) [inline slamon_afm/models.py]: Your way is better tho :+1:
